File: clients/swift/Sources/messaging-cli/MessageClient.swift
import Foundation

actor MessageClient {
    enum State: String {
        case loggedOut
        case registering
        case loggingIn
        case online
        case flushing
        case offline
    }

    var serverURL: String
    var onMessageReceived: ((WireMessage) -> Void)?
    var onStateChange: ((String) -> Void)?
    var onError: ((String) -> Void)?

    private let network = NetworkClient()
    private let queue = OfflineQueue()
    private var state: State = .loggedOut
    private var sessionId: String?
    private var currentUser: String?
    private var lastServerTimestamp: Int64 = 0
    private var syncTask: Task<Void, Never>?
    let networkMonitor = NetworkMonitor()

    init(serverURL: String) {
        self.serverURL = serverURL
    }

    func setCallbacks(
        onMessageReceived: @escaping (WireMessage) -> Void,
        onStateChange: @escaping (String) -> Void,
        onError: @escaping (String) -> Void
    ) {
        self.onMessageReceived = onMessageReceived
        self.onStateChange = onStateChange
        self.onError = onError
    }

    func setupNetworkMonitor() {
        networkMonitor.onStatusChange = { [weak self] isOnline in
            Task { [weak self] in
                await self?.handleNetworkChange(isOnline: isOnline)
            }
        }
    }

    private func handleNetworkChange(isOnline: Bool) async {
        if isOnline {
            guard state == .offline else { return }
            if queue.nextPending() != nil {
                await transition(to: .flushing)
                await flushQueue()
            } else {
                await transition(to: .online)
                startSyncLoop(immediate: true)
            }
        } else {
            if state == .online || state == .flushing {
                syncTask?.cancel()
                syncTask = nil
                await transition(to: .offline)
            }
        }
    }

    func register(username: String, password: String) async {
        guard state == .loggedOut else { return }
        await transition(to: .registering)
        do {
            _ = try await network.register(username: username, password: password, serverURL: serverURL)
            await transition(to: .loggedOut)
        } catch NetworkError.usernameTaken {
            onError?("username_taken")
            await transition(to: .loggedOut)
        } catch {
            onError?("registration_failed: \(error)")
            await transition(to: .loggedOut)
        }
    }

    func login(username: String, password: String) async {
        guard state == .loggedOut else { return }
        await transition(to: .loggingIn)
        do {
            let resp = try await network.login(username: username, password: password, serverURL: serverURL)
            sessionId = resp.sessionId
            currentUser = resp.username
            await transition(to: .online)
            startSyncLoop(immediate: true)
        } catch NetworkError.unauthorized {
            onError?("invalid_credentials")
            await transition(to: .loggedOut)
        } catch {
            onError?("login_failed: \(error)")
            await transition(to: .loggedOut)
        }
    }

    func sendMessage(to: String, text: String) async {
        let localId = queue.enqueue(to: to, text: text)

        switch state {
        case .online:
            guard let sid = sessionId else { return }
            queue.markSending(localId: localId)
            do {
                let resp = try await network.send(to: to, text: text, sessionId: sid, serverURL: serverURL)
                queue.markSent(localId: localId, serverId: resp.id)
            } catch NetworkError.unauthorized {
                queue.markPending(localId: localId)
                await handleUnauthorized()
            } catch NetworkError.connectionFailed {
                queue.markPending(localId: localId)
                syncTask?.cancel()
                syncTask = nil
                await transition(to: .flushing)
                await transition(to: .offline)
            } catch {
                queue.markPending(localId: localId)
                syncTask?.cancel()
                syncTask = nil
                await transition(to: .flushing)
                await flushQueue()
            }
        case .flushing, .offline:
            // Already queued above; nothing else to do
            break
        default:
            break
        }
    }

    func logout() async {
        if let sid = sessionId {
            await network.logout(sessionId: sid, serverURL: serverURL)
        }
        sessionId = nil
        currentUser = nil
        syncTask?.cancel()
        syncTask = nil
        await transition(to: .loggedOut)
    }

    private func transition(to newState: State) async {
        state = newState
        onStateChange?(newState.rawValue)
    }

    private func startSyncLoop(immediate: Bool) {
        syncTask?.cancel()
        syncTask = Task {
            if immediate {
                await pollMessages()
            }
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 3_000_000_000)
                guard !Task.isCancelled else { break }
                await pollMessages()
            }
        }
    }

    private func pollMessages() async {
        guard let sid = sessionId, state == .online else { return }
        do {
            let resp = try await network.getMessages(since: lastServerTimestamp, sessionId: sid, serverURL: serverURL)
            lastServerTimestamp = resp.serverTimestamp
            for msg in resp.messages {
                if let serverId = msg.id as String?, queue.isDuplicate(serverId: serverId) { continue }
                onMessageReceived?(msg)
            }
        } catch NetworkError.unauthorized {
            await handleUnauthorized()
        } catch NetworkError.connectionFailed {
            syncTask?.cancel()
            syncTask = nil
            await transition(to: .offline)
        } catch {
            // transient error — continue loop
        }
    }

    private func flushQueue() async {
        guard state == .flushing, let sid = sessionId else { return }

        while true {
            guard let msg = queue.nextPending() else {
                // Queue drained
                await transition(to: .online)
                startSyncLoop(immediate: true)
                return
            }

            queue.markSending(localId: msg.localId)

            do {
                let resp = try await network.send(to: msg.toUser, text: msg.text, sessionId: sid, serverURL: serverURL)
                queue.markSent(localId: msg.localId, serverId: resp.id)
            } catch NetworkError.unauthorized {
                queue.markPending(localId: msg.localId)
                await handleUnauthorized()
                return
            } catch NetworkError.connectionFailed {
                queue.markPending(localId: msg.localId)
                await transition(to: .offline)
                return
            } catch NetworkError.serverError(let code) where code >= 500 {
                queue.markFailed(localId: msg.localId)
                // continue to next message
            } catch {
                queue.markFailed(localId: msg.localId)
                // continue to next message
            }
        }
    }

    private func handleUnauthorized() async {
        sessionId = nil
        currentUser = nil
        syncTask?.cancel()
        syncTask = nil
        await transition(to: .loggedOut)
    }
}

File: clients/kotlin/src/main/kotlin/com/messaging/MessageClient.kt
package com.messaging

import kotlinx.coroutines.*

enum class ClientState { LOGGED_OUT, REGISTERING, LOGGING_IN, ONLINE, FLUSHING, OFFLINE }

class MessageClient(
    private val serverURL: String,
    private val networkMonitor: NetworkMonitor,
    private val networkClient: NetworkClient = NetworkClient(),
    private val offlineQueue: OfflineQueue = OfflineQueue()
) {
    var onMessageReceived: ((WireMessage) -> Unit)? = null
    var onStateChange: ((String) -> Unit)? = null
    var onError: ((String) -> Unit)? = null

    @Volatile var state: ClientState = ClientState.LOGGED_OUT
    @Volatile private var sessionId: String? = null
    @Volatile private var lastServerTimestamp: Long = 0L
    private var currentJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    init {
        networkMonitor.onStatusChange = { isOnline ->
            scope.launch {
                if (!isOnline) {
                    if (state == ClientState.ONLINE || state == ClientState.FLUSHING) {
                        currentJob?.cancel()
                        changeState(ClientState.OFFLINE)
                    }
                } else {
                    if (state == ClientState.OFFLINE) {
                        if (offlineQueue.nextPending() != null) {
                            changeState(ClientState.FLUSHING)
                            currentJob = scope.launch { runFlushQueue() }
                        } else {
                            changeState(ClientState.ONLINE)
                            startSyncLoop()
                        }
                    }
                }
            }
        }
    }

    private fun changeState(newState: ClientState) {
        state = newState
        onStateChange?.invoke(newState.name)
    }

    suspend fun register(username: String, password: String) {
        changeState(ClientState.REGISTERING)
        try {
            networkClient.register(username, password, serverURL)
            changeState(ClientState.LOGGED_OUT)
        } catch (e: NetworkError.UsernameTaken) {
            onError?.invoke("username_taken")
            changeState(ClientState.LOGGED_OUT)
        } catch (e: NetworkError) {
            onError?.invoke("connection_failed")
            changeState(ClientState.LOGGED_OUT)
        }
    }

    suspend fun login(username: String, password: String) {
        changeState(ClientState.LOGGING_IN)
        try {
            val response = networkClient.login(username, password, serverURL)
            sessionId = response.sessionId
            offlineQueue.initialize(username)
            changeState(ClientState.ONLINE)
            startSyncLoop()
        } catch (e: NetworkError.Unauthorized) {
            onError?.invoke("invalid_credentials")
            changeState(ClientState.LOGGED_OUT)
        } catch (e: NetworkError) {
            onError?.invoke("connection_failed")
            changeState(ClientState.LOGGED_OUT)
        }
    }

    suspend fun sendMessage(to: String, text: String) {
        val localId = offlineQueue.enqueue(to, text)
        if (state == ClientState.ONLINE) {
            val sid = sessionId
            if (sid == null) {
                offlineQueue.markPending(localId)
                return
            }
            offlineQueue.markSending(localId)
            try {
                val response = networkClient.send(to, text, sid, serverURL)
                offlineQueue.markSent(localId, response.id)
            } catch (e: NetworkError.Unauthorized) {
                offlineQueue.markPending(localId)
                sessionId = null
                currentJob?.cancel()
                changeState(ClientState.LOGGED_OUT)
            } catch (e: NetworkError.ConnectionFailed) {
                offlineQueue.markPending(localId)
                currentJob?.cancel()
                changeState(ClientState.OFFLINE)
            } catch (e: NetworkError) {
                offlineQueue.markFailed(localId)
                changeState(ClientState.FLUSHING)
                currentJob = scope.launch { runFlushQueue() }
            }
        }
        // OFFLINE or FLUSHING: already enqueued, nothing more to do
    }

    suspend fun logout() {
        try {
            sessionId?.let { networkClient.logout(it, serverURL) }
        } catch (e: Exception) {
            // best-effort
        }
        sessionId = null
        currentJob?.cancel()
        changeState(ClientState.LOGGED_OUT)
    }

    private fun startSyncLoop() {
        currentJob?.cancel()
        currentJob = scope.launch {
            // immediate first poll on transition to ONLINE
            poll()
            while (isActive && state == ClientState.ONLINE) {
                delay(3000)
                if (state == ClientState.ONLINE) poll()
            }
        }
    }

    private suspend fun poll() {
        val sid = sessionId ?: return
        try {
            val response = networkClient.getMessages(lastServerTimestamp, sid, serverURL)
            lastServerTimestamp = response.serverTimestamp
            response.messages.forEach { msg ->
                if (!offlineQueue.isDuplicate(msg.id)) {
                    onMessageReceived?.invoke(msg)
                }
            }
        } catch (e: NetworkError.Unauthorized) {
            sessionId = null
            changeState(ClientState.LOGGED_OUT)
        } catch (e: NetworkError.ConnectionFailed) {
            changeState(ClientState.OFFLINE)
        } catch (e: NetworkError) {
            // transient error — continue
        }
    }

    private suspend fun runFlushQueue() {
        while (state == ClientState.FLUSHING) {
            val msg = offlineQueue.nextPending() ?: break
            val sid = sessionId
            if (sid == null) {
                changeState(ClientState.LOGGED_OUT)
                return
            }
            offlineQueue.markSending(msg.localId)
            try {
                val response = networkClient.send(msg.toUser, msg.text, sid, serverURL)
                offlineQueue.markSent(msg.localId, response.id)
            } catch (e: NetworkError.Unauthorized) {
                offlineQueue.markPending(msg.localId)
                sessionId = null
                changeState(ClientState.LOGGED_OUT)
                return
            } catch (e: NetworkError.ConnectionFailed) {
                offlineQueue.markPending(msg.localId)
                changeState(ClientState.OFFLINE)
                return
            } catch (e: NetworkError.ServerError) {
                offlineQueue.markFailed(msg.localId)
                // 5xx: log and continue to next message
            } catch (e: NetworkError) {
                offlineQueue.markFailed(msg.localId)
            }
        }
        if (state == ClientState.FLUSHING) {
            changeState(ClientState.ONLINE)
            startSyncLoop()
        }
    }
}
